### ISSUE TITLE
Support `root:` event target in `@HostListener`

### ISSUE THREAD

**Issue by MillerSvt** [angular/angular#70150]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

`@HostListener` supports global event targets such as `document:`, `window:` and `body:`:

```ts
@HostListener('document:click', ['$event'])
onClick(event: MouseEvent): void {
    // ...
}
```

There is currently no equivalent target for the root node of the directive/component host element (`hostElement.getRootNode()`).

This becomes important for directives and components that can run both in the regular document and inside Shadow DOM. Listening on `document` is not equivalent to listening on the current root: when an element is rendered inside a `ShadowRoot`, the natural event scope can be that shadow root rather than the owner document.

Today this requires dropping the declarative `@HostListener` API and manually resolving the root and managing the listener lifecycle, for example:

```ts
private readonly element = inject(ElementRef<HTMLElement>);
private readonly destroyRef = inject(DestroyRef);

constructor() {
    const root = this.element.nativeElement.getRootNode();
    const listener = (event: Event) => this.onClick(event);

    root.addEventListener('click', listener);
    this.destroyRef.onDestroy(() => root.removeEventListener('click', listener));
}

private onClick(event: Event): void {
    // ...
}
```

### Proposed solution

Add a `root:` event target which resolves to `getRootNode()` of the host element:

```ts
@HostListener('root:click', ['$event'])
onClick(event: MouseEvent): void {
    // ...
}
```

Conceptually, Angular would resolve `root:` at directive/component creation time as:

```ts
hostElement.getRootNode()
```

This means:

- for a regular element in the document, the target is `Document`;
- for an element inside Shadow DOM, the target is its `ShadowRoot`.

It would make listeners naturally follow the DOM tree in which the host is actually rendered and would provide a declarative counterpart to manually calling `getRootNode().addEventListener(...)`.

If the same event-target syntax is shared with template event bindings, supporting `(root:click)` there as well would also be useful and consistent.

### Alternatives considered

**1. `@HostListener('document:click')`**

This does not express the same semantics. A reusable directive should not have to know whether its host lives directly in `Document` or in a `ShadowRoot`.

**2. Manual `addEventListener` / `removeEventListener`**

This works, but requires resolving `getRootNode()`, retaining the listener reference, and handling destruction manually. It loses the concise declarative lifecycle management provided by `@HostListener`.

**3. RxJS `fromEvent(hostElement.getRootNode(), 'click')`**

This also works, but similarly requires imperative setup and teardown (`takeUntilDestroyed`, etc.) for something that otherwise maps naturally to `@HostListener`.

**4. A custom `EventManagerPlugin`**

Applications can implement their own event target syntax, but this is disproportionate for a standard DOM concept (`Node.getRootNode()`) and makes reusable libraries depend on application-level event manager configuration.

### Additional

If this approach is approved, I’d be happy to submit a PR with the implementation and tests.

**Comment by MillerSvt** [comment]
It may also be worth considering, separately from `root:`, whether Angular should support a few more well-defined browser-level `EventTarget`s, such as `visualViewport` and `screen.orientation`.

For example:

```ts
@HostListener('visualViewport:resize', ['$event'])
@HostListener('orientation:change', ['$event'])
```

I would probably keep this as a separate consideration from `root:`, since `root:` is directly derived from the host element via `getRootNode()`, while these are additional global browser targets.

**Comment by MillerSvt** [comment]
After looking at the current listener lifecycle more closely, there is one important architectural complication with `root:` that is probably worth calling out explicitly.

Today a global target resolver is evaluated when the listener is registered:

```ts
const native = getNativeByTNode(tNode, lView) as RElement;
const target = eventTargetResolver ? eventTargetResolver(native) : native;
const cleanupFn = renderer.listen(target as RElement, eventName, wrappedListener);
```

The teardown function is then stored in the view cleanup data and normally runs when the view is destroyed. In other words, the listener lifecycle is currently tied to Angular view creation/destruction, not to the physical DOM attachment/location of the host node.

That is fine for the existing targets because `window`, `document`, and `body` are effectively stable for normal Angular DOM moves. `getRootNode()`, however, is location-dependent.

A naive resolver such as:

```ts
export function ɵɵresolveRoot(element: RElement): EventTarget {
  return element.getRootNode();
}
```

would therefore only capture the root that happens to exist at listener-registration time.

### Case 1: the node is not attached yet

A DOM node can exist before it is attached to its final DOM tree. For a detached element:

```ts
const element = document.createElement('div');

element.getRootNode() === element;
```

So if the `root:` listener is registered before the node is inserted into a `Document` or `ShadowRoot`, Angular could end up subscribing to the element itself rather than to its eventual root.

This can be relevant for dynamically created/detached views and components, where the Angular view can exist before its DOM nodes are inserted into their final location.

### Case 1.2: the node may belong to a detached subtree during listener registration

The host element does not have to be completely standalone when the listener is registered. It may already be attached to a parent, while that whole subtree is still detached from the final DOM tree:

```ts
const wrapper = document.createElement('div');
const element = document.createElement('button');

wrapper.append(element);

element.getRootNode() === wrapper;
```

If Angular registers the listener at this point, `root:` would resolve to `wrapper`.

Later the same subtree may be inserted into a shadow root:

```ts
shadowRoot.append(wrapper);

element.getRootNode() === shadowRoot;
```

So the initial problem is broader than `getRootNode()` returning the element itself: it can return any temporary detached tree root that is not the element's eventual `Document` or `ShadowRoot`.

### Case 2: the root can change during the lifetime of the same node/view

More importantly, the root of an existing native node is not necessarily stable.

The node, its directive instance, and its `LView` may all remain alive while Angular changes the node's physical DOM location. For example:

```text
shadowRootA
  └── element

      ↓ move

shadowRootB
  └── element
```

Before the move:

```ts
element.getRootNode() === shadowRootA;
```

After the move:

```ts
element.getRootNode() === shadowRootB;
```

However, the listener created earlier by:

```ts
const target = eventTargetResolver(native);
const cleanupFn = renderer.listen(target, eventName, wrappedListener);
```

would still be attached to `shadowRootA`.

There are multiple Angular mechanisms that can result in this kind of physical DOM relocation without necessarily destroying and recreating the corresponding directive/view:

- content projection, including changing where projected content is rendered;
- detaching and reinserting an existing view, e.g. via `ViewContainerRef.detach()` / `insert()`;
- moved/transplanted embedded views;
- other Angular-controlled insert/detach/move operations.

There is also an intermediate detached state to consider:

```text
shadowRootA
  └── element

      ↓ detach

element.getRootNode() === element

      ↓ insert

shadowRootB
  └── element
```

If `root:` is intended to mean "the current root", the expected listener lifecycle is therefore not just creation/destruction. It is closer to:

```text
insert into shadowRootA
  -> subscribe to shadowRootA

detach
  -> unsubscribe from shadowRootA

insert into shadowRootB
  -> subscribe to shadowRootB

destroy
  -> final cleanup
```

This is the main difference between `root:` and the existing `window:`, `document:` and `body:` targets: those targets are effectively stable across normal Angular DOM relocation, while `getRootNode()` is inherently dependent on the node's current physical location.

### Possible implication for the implementation

Because of the above, `root:` is somewhat different from the current `GlobalTargetResolver` use cases. The resolver API already receives the host element, but it is currently evaluated as part of listener registration and is not a reactive/dynamic binding to the node's location.

A fully correct implementation may require some notion of a location-dependent event target, for example:

```ts
ɵɵlistener('click', handler, ɵɵresolveRoot, /* dynamic target */ true);
```

or an equivalent internal flag/metadata so Angular can:

```text
attach node
  -> resolve getRootNode()
  -> subscribe

move node
  -> unsubscribe old root
  -> resolve new root
  -> subscribe

detach node
  -> unsubscribe

destroy view
  -> final cleanup
```

Ideally this would only add work for listeners whose target actually depends on node location; doing this for all listeners on every DOM insertion/removal would probably be unnecessary overhead.

There is one interesting detail in the current cleanup metadata: for global-target listeners Angular stores a target getter based on the resolver:

```ts
const idxOrTargetGetter = eventTargetResolver
  ? (_lView: LView) => eventTargetResolver(unwrapRNode(_lView[tNode.index]))
  : tNode.index;
```

so the infrastructure already retains enough information to recompute the target.

However, the actual `cleanupFn` returned by `renderer.listen(...)` still closes over the original subscription target, so this alone does not provide re-binding.

This makes `root:` a bit more than just adding `resolveRoot` to `GLOBAL_TARGET_RESOLVERS`. The public API is simple, but the desired semantics probably need to be agreed on first:

- should `root:` mean the root at listener-registration time;
- or the current root of the host element?

I think users would naturally expect the latter.

**Comment by Deep-Axe** [comment]
`root:` can either have snapshot or placement-aware semantics.

A fourth `GLOBAL_TARGET_RESOLVERS` entry would snapshot `host.getRootNode()` when the listener is registered. Unlike ```window:, document:, and body:```, that target can change when Angular inserts, detaches, projects, or moves the host. The listener cleanup also closes over the original target, so the stored target getter does not provide rebinding.

This is how the contract could look like
  > - If the host’s current root is a ```Document or ShadowRoot```, listen on that root.
  > - If the host is detached, keep the listener dormant rather than listening on the detached element or ancestor
  >   returned by ```getRootNode()```.
  >
  > - After an Angular-managed insertion, move, or reprojection, bind or rebind if the root changed.
  > - On detach, unsubscribe.
  > - On destroy, perform final cleanup.

This would need a dedicated location-dependent listener kind/state, instead of  another `ɵɵresolveWindow`-styleresolver The integration point needs to be thought of: creation does not always use the insert walk, projection uses ```WalkTNodeTreeAction.Create```, and moving an ancestor does not visit every listener-bearing descendant or child component view.

 ```(root:click)``` in templates should be supported alongside ```@HostListener('root:click')```, since both share the compiler target machinery.

We should also account for
  > - SSR renderers where ```getRootNode()``` is unavailable;
  > - compiler/linker recognition of the root target (today it is rejected as an unknown global target)
  > - partial-compilation ```minVersion```, so libraries using root: are not linked by an older linkr that cannot
  >   understand it.

Tests could start with light DOM, native Shadow DOM, an initially detached ```TemplateRef``` view inserted through ```ViewContainerRef```, detach/reinsert across roots, nested projected content in ```ViewEncapsulation.ShadowDom```, component-host moves, ng-container/ng-template hosts, SSR, and destruction while detached.

If placement-aware lifecycle support is more than we want in core, it would be smarter not to add root: since we would end up exposing snapshot semantics that are easy to misuse.
